docs/book/tutorials/kubernetes-on-vsphere-with-kubeadm.md

```diff
@@ -156,262 +213,996 @@ bootstrapTokens:
+```
+# tee /etc/kubernetes/kubeadminitworker.yaml >/dev/null <<EOF
 apiVersion: kubeadm.k8s.io/v1beta1
 caCertPath: /etc/kubernetes/pki/ca.crt
 discovery:
   file:
-    kubeConfigPath: discovery.yaml
+    kubeConfigPath: /etc/kubernetes/discovery.yaml
   timeout: 5m0s
   tlsBootstrapToken: y7yaev.9dvwxx6ny4ef8vlq
 kind: JoinConfiguration
 nodeRegistration:
   criSocket: /var/run/dockershim.sock
   kubeletExtraArgs:
-    cloud-provider: vsphere
+    cloud-provider: external
+EOF
+```
+
+You can copy the `discovery.yaml` to your local machine with `scp`. In the example below, we are first copying it locally to a central desktop, and then copying it out to all the nodes.
+```
+# scp ubuntu@10.192.116.47:~/discovery.yaml discovery.yaml
+
+# scp discovery.yaml ubuntu@10.192.116.46:~/discovery.yaml
+# scp discovery.yaml ubuntu@10.192.116.48:~/discovery.yaml
+# scp discovery.yaml ubuntu@10.192.116.49:~/discovery.yaml
+# scp discovery.yaml ubuntu@10.192.116.50:~/discovery.yaml
+```
+You will now need to login to each of the nodes and copy the `discovery.yaml` file from `/home/ubuntu` to `/etc/kubernetes`. The `discovery.yaml` file must exist in `/etc/kubernetes/discovery.yaml`.
+Once that step is completed, run the following command on each worker node to have that worker node join the master (and other worker nodes) in the cluster:
+```
+# kubeadm join --config /etc/kubernetes/kubeadminitworker.yaml
+```
+
+## Install the vSphere Cloud Provider Interface
+The following steps are only done on the master. Please note that the CSI driver requires the presence of the Cloud Provider Interface (CPI), so this step of installing the CPI is mandatory.
+
+### Create a CPI configMap ###
+This cloud-config configmap file, passed to the CPI on initialization, contains details about the vSphere configuration. This file, which here we have called `vsphere.conf` has been populated with some sample values. Obviously, you will need to modify this file to reflect your own vSphere configuration.
+```
+# tee /etc/kubernetes/vsphere.conf >/dev/null <<EOF
+[Global]
+insecure-flag = "true"
+  
+[VirtualCenter "1.1.1.1"]
+user = "Administrator@vsphere.local"
+password = "Admin!23"
+port = "443"
+datacenters = "datacenter"
+  
+[Network]
+public-network = "VM Network"
 EOF
 ```
+Here is a description of the fields used in the vsphere.conf configmap.
+* `insecure-flag` should be set to true to use self-signed certificate for login
+* `VirtualCenter` section is defined to hold property of vcenter. IP address and FQDN should be specified here.
+* `user` is the vCenter username for vSphere Cloud Provider.
+* `password` is the password for vCenter user specified with user.
```

Comment:
We aren't documenting using secrets? I think we should remove the documentation for setting username/password in the configmap in favor of using k8s secrets.

Reply:
We took this procedure from the CSI/CNS team. I'll work with Myles on the secrets concern. Though we might just merge this, and open a new PR to address this.